src/kmp.h

```diff
@@ -0,0 +1,68 @@
+#ifndef KMP_H
+#define KMP_H
+
+#include <malloc.h>
+
+int *compute_border_array(char *pattern, int pattern_len) {
+    if (pattern_len == 0) return NULL;
+    int *ba = malloc(sizeof(int)*pattern_len);
```

Comment:
Here's a pattern you should get used to in C:

```c
int *ba = malloc(pattern_len * sizeof *ba);
```

I didn't use it in the book because I didn't want to go into technical details of C, but if you plan to learn the language, you want to stick to this pattern. (I regret my choice of not using it in the book on a daily basis; I think it is one of the cases where simplifying does more damage than good).

The `sizeof` keyword can take a type as an argument in two ways: an actual type (in which case you need parentheses) or a variable that has that type (in which case you don't need parentheses, but you can have them if you want; I usually leave them out to make the distinction clear).

If you allocate memory to hold a given type

```c
type *var = malloc(sizeof(type));
```

it is all too easy to change the type at some point and end up in a situation like this:

```c
type2 *var = malloc(sizeof(type));
```

You have changed the type of the pointer, but you forgot to update how much memory you allocated.

If you use

```c
type *var = malloc(sizeof *var)
```

you are telling `sizeof` that you want the size of what `var` is pointing at (don't worry, you are not dereferencing anything here, you are just computing a type). That ensures that you get a chunk of memory that matches what `var` will point to.

The allocation in your code is only a tad more complicated because you allocate more than one `int` (you allocate `pattern_len` of them), but the principle is the same: `sizeof *ba` is the size of whatever it is `ba` points to (here `int`).